File: MATLAB/powerhead.m
% Lox Pump, Fuel Pump (each with impeller/blades/volute/inducer), Turbine, Gas Generator
% Lox pump: p=1; Fuel pump: p=2

%% Total Mass Flow Rate INITIAL GUESSES
mdot_gg = mdot_cc*gg_fraction_guess/(1-gg_fraction_guess); % kg/s - Propellant mass flow rate into gas generator
mdot_fuel_gg = mdot_gg*(1/(1+OF_gg)); % kg/s - Fuel Mass Flow Rate
mdot_ox_gg = mdot_gg*(OF_gg/(1+OF_gg)); % kg/s - Oxidizer Mass Flow Rate
mdot_fuel_total = mdot_fuel_cc+mdot_fuel_gg; % kg/s
mdot_ox_total = mdot_ox_cc+mdot_ox_gg; % kg/s
% actual mdot_gg is not solved until turbine is called

%% Liquid Oxygen Pump
p_out(1) = p_cc*(1+cc_stiffness) + 10*6894.76; % Pa - injector inlet pressure plus 10 psi of margin for plumbing losses
p_in(1) = p_amb; % THIS ISN'T TRUE DOWNSTREAM OF INDUCER - Pa - inlet pressure; assumes zero dp across inducer for conservatism 
T_in(1) = PropsSI("T", "P", p_amb, "Q", 0, "Oxygen");
density_ox = PropsSI('D','Q',0,'P',p_in(1),'Oxygen'); % kg/m3 - Lox at saturation
density_pump(1) = density_ox; % kg/m3
mdot_pump(1) = mdot_ox_total; % kg/s - Oxidizer Mass Flow Rate
vapor_pressure_pump(1) = PropsSI("P", "T", T_in(1), "Q", 0, "Oxygen");
clock(1) = 1; % Counterclockwise
hub_tip_ratio_inducer(1) = 0.3; % unitless - between 0.2 and 0.4 for rear drive and between 0.5 and 0.6 for front drive

p=1; % sets the 'active' pump to the ox pump
pump

    writematrix([impeller_curve(:,:,p)*1000, zeros(length(impeller_curve(:,1)),1)], 'Curves/ox_impeller_curve_mm.txt', 'Delimiter', ',')  
    writematrix([shroud_curve(:,:,p)*1000, zeros(length(shroud_curve(:,1)),1)], 'Curves/ox_shroud_curve_mm.txt', 'Delimiter', ',')
    writematrix([blade_curve(:,1,p)*1000, zeros(length(impeller_curve(:,1)),1), -blade_curve(:,2,p)*1000], 'Curves/ox_blade_curve_mm.txt', 'Delimiter', ',')
    writematrix([volute_curve(:,1,p)*1000, volute_curve(:,2,p)*1000, zeros(length(volute_curve(:,1,p)),1)], 'Curves/ox_volute_curve_mm.txt', 'Delimiter', ',')
    
    data_labels = {'h_volute_ox';'t_tongue_ox';'r_volute_outlet_ox';'OD_volute_outlet_ox'};
    data = table(data_labels, {h_volute(p)*1000;t_tongue(p)*1000;r_volute_outlet(p)*1000;OD_volute_outlet(p)*1000});
    writetable(data, 'Output Parameters/ox_volute_mm.txt', 'WriteVariableNames',0);
    
    data_labels = {'h_min_ox';'r_tip_ox';'r_hub_base_ox';'r_hub_top_ox';'blade_lead_ox';'clearance_ox'};
    data = table(data_labels, {h_min_inducer(p)*1000;r_tip_inducer(p)*1000;r_hub_base(p)*1000;r_hub_top(p)*1000;blade_lead_inducer(p)*1000;clearance_axial_inducer(p)*1000});
    writetable(data, 'Output Parameters/ox_inducer_mm.txt', 'WriteVariableNames',0);

%% Fuel Pump
p_out(2) = p_cool(1) + 10*6894.76; % Pa - regen inlet pressure plus 10 psi of margin for plumbing losses
p_in(2) = p_amb; % THIS ISN'T TRUE DOWNSTREAM OF INDUCER - Pa - inlet pressure; assumes zero dp across inducer for conservatism
T_in(2) = T_amb; % K
density_fuel = 1000*double(rocketprop('RP1').SGLiqAtTdegR(T_in(2)*1.8)); % kg/m3 - Jet-A at ambient temperature
density_pump(2) = 1000*double(rocketprop('RP1').SGLiqAtTdegR(T_in(2)*1.8)); % kg/m3
mdot_pump(2) = mdot_fuel_total; % kg/s - Fuel Mass Flow Rate
vapor_pressure_pump(2) = 0.001*double(rocketprop('RP1').PvapAtTdegR(T_amb*1.8)); % kg/m3 - ethanol at STP
clock(2) = -1;
hub_tip_ratio_inducer(2) = 0.3; % unitless - between 0.2 and 0.4 for rear drive and between 0.5 and 0.6 for front drive

p=2; % sets the 'active' pump to the fuel pump
pump

    writematrix([impeller_curve(:,:,p)*1000, zeros(length(impeller_curve(:,1)),1)], 'Curves/fuel_impeller_curve_mm.txt', 'Delimiter', ',')  
    writematrix([shroud_curve(:,:,p)*1000, zeros(length(shroud_curve(:,1)),1)], 'Curves/fuel_shroud_curve_mm.txt', 'Delimiter', ',')
    writematrix([blade_curve(:,1,p)*1000, zeros(length(impeller_curve(:,1)),1), -blade_curve(:,2,p)*1000], 'Curves/fuel_blade_curve_mm.txt', 'Delimiter', ',')
    writematrix([volute_curve(:,1,p)*1000, volute_curve(:,2)*1000, zeros(length(volute_curve(:,1,p)),1)], 'Curves/fuel_volute_curve_mm.txt', 'Delimiter', ',')
    
    data_labels = {'h_volute_fuel';'t_tongue_fuel';'r_volute_outlet_fuel';'OD_volute_outlet_fuel'};
    data = table(data_labels, {h_volute(p)*1000;t_tongue(p)*1000;r_volute_outlet(p)*1000;OD_volute_outlet(p)*1000});
    writetable(data, 'Output Parameters/fuel_volute_mm.txt', 'WriteVariableNames',0);
    
    data_labels = {'h_min_fuel';'r_tip_fuel';'r_hub_base_fuel';'r_hub_top_fuel';'blade_lead_fuel';'clearance_fuel'};
    data = table(data_labels, {h_min_inducer(p)*1000;r_tip_inducer(p)*1000;r_hub_base(p)*1000;r_hub_top(p)*1000;blade_lead_inducer(p)*1000;clearance_axial_inducer(p)*1000});
    writetable(data, 'Output Parameters/fuel_inducer_mm.txt', 'WriteVariableNames',0);

%% Gas Generator
gas_generator

%% Turbine
shaft_power_pump = sum(power_pump);
shaft_power_pump_theory = sum(power_pump_theory);
shaft_power_turbine = shaft_power_pump/gear_efficiency; % unitless - the power through the turbine SHAFT - not the turbine itself, which will be subject to a stage inefficiency

turbine

    % Write results to text file
    data_labels = {'Base Radius (mm)'; 'Pitchline Radius (mm)'; 'Tip Radius (mm)'; 'Blade Width (mm)'; 'Blade Depth (mm)';
        'Blade Angle (deg)'; 'Blade Opening (mm)'; 'Blade Number'; 'Nozzle Angle (deg)'; 'Nozzle Diameter (mm)'; 'Nozzle Number'};
    data = table(data_labels, {r_base_rotor*1000; r_pitchline_rotor*1000; r_tip_rotor*1000; blade_width_rotor*1000; blade_length_rotor*1000; ...
        incidence_angle_blade*180/pi; blade_opening_rotor*1000; blade_count_rotor; angle_nozzle*180/pi; d_throat_nozzle*1000; nozzle_number});
    writetable(data, 'Output Parameters/turbine_mm.txt', 'WriteVariableNames',0);

% Efficiency
transmission_efficiency = shaft_power_pump_theory / shaft_power_turbine; % unitless
isp_real = thrust / mdot_total / g; % s

% Shaft
shaft_torque_pump = shaft_power_pump./abs(pump_shaft_speed); % N*m
shaft_polar_moment = pi/2*r_shaft^4; % m4
shaft_stress = shaft_torque_pump*r_shaft/shaft_polar_moment; % Pa
shaft_torsion_FSu = shaft_shear_strength ./ shaft_stress; % unitless
shaft_torsion_MSu = shaft_torsion_FSu / FSu_design - 1; % unitless

File: Python/powerhead.py
from dataclasses import dataclass
import numpy as np
from engine_state import engine, tca, gg, pump, turbine

from pumps import pumps
from gas_generator import gas_generator
from turbines import turbines

from CoolProp.CoolProp import PropsSI
from rocketprops.rocket_prop import get_prop as rprop

def powerhead():

    rprop_fuel = rprop(engine.fuelrp)
    rprop_oxidizer = rprop(engine.oxidizer)

    # Total Mass Flow Rate INITIAL GUESSES
    gg.mdot = tca.mdot*gg.fraction_guess/(1-gg.fraction_guess)       # kg/s - mass flow rate through gas generator
    gg.mdot_fuel = gg.mdot*(1/(1+gg.mixture_ratio))                             # kg/s - fuel mass flow rate through gas generator
    gg.mdot_ox   = gg.mdot*(gg.mixture_ratio/(1+gg.mixture_ratio))                         # kg/s - oxidizer mass flow rate through gas generator
    engine.mdot_fuel_total = tca.mdot_fuel + gg.mdot_fuel            # kg/s - total fuel mass flow rate
    engine.mdot_ox_total   = tca.mdot_ox   + gg.mdot_ox              # kg/s - total oxidizer mass flow rate
    # TODO: add a verification and/or iteration on the gg fraction guess at end of powerhead module

    # Liquid Oxygen Pump
    ox_pump = pump(fluid='ox')
    ox_pump.p_out = tca.p_injector_manifold + 10*6894.76                    # Pa - add 10 psi of margin for plumbing losses
    ox_pump.p_in  = ox_pump.tank.p                                           # Pa   
    ox_pump.T_in  = rprop_oxidizer.TdegRAtPsat(engine.p_amb/6894.76)/1.8     # K - tank temperature - assumes saturated propellant during fill
    ox_pump.density = 1000*rprop_oxidizer.SGLiqAtTdegR(ox_pump.T_in*1.8)      # kg/m3 - density of LOX at inlet
    ox_pump.viscosity = 0.1*rprop_oxidizer.ViscAtTdegR(ox_pump.T_in*1.8)     # Pa-s - viscosity of LOX at inlet
    ox_pump.mdot = engine.mdot_ox_total                                      # kg/s     
    ox_pump.pvap_inlet = 6894.76*rprop_oxidizer.PvapAtTdegR(ox_pump.T_in*1.8)        # Pa - vapor pressure of LOX at tank temperature
  
    # Fuel Pump
    fuel_pump = pump(fluid='fuel')
    fuel_pump.p_out = tca.regenerative_coolant.p[tca.regenerative_coolant.flow_direction] + fuel_pump.plumbing_loss # add margin for plumbing losses                   # Pa
    fuel_pump.p_in = fuel_pump.tank.p                                                       # Pa
    fuel_pump.T_in = engine.T_amb                                                       # K - tank temperature    
    fuel_pump.density   = 1000*rprop_fuel.SGLiqAtTdegR(fuel_pump.T_in*1.8)      # kg/m3 - density of fuel at inlet
    fuel_pump.viscosity = 0.1*rprop_fuel.ViscAtTdegR(fuel_pump.T_in*1.8)     # Pa-s - viscosity of fuel at inlet
    fuel_pump.mdot  = engine.mdot_fuel_total                                            # kg/s
    fuel_pump.pvap_inlet = 6894.76*rprop_fuel.PvapAtTdegR(fuel_pump.T_in*1.8) # Pa - vapor pressure of RP1 at tank temperature
    
    # Find Shaft Speed
    ox_pump.shaft_speed = fuel_pump.shaft_speed = None
    pumps(ox_pump)
    pumps(fuel_pump)

    ox_pump.shaft_speed = fuel_pump.shaft_speed = np.min([ox_pump.shaft_speed, fuel_pump.shaft_speed, engine.shaft_speed_cap]) # rad/s

    # Actual Runs
    pumps(ox_pump)
    pumps(fuel_pump)

    gg.pc = np.min([fuel_pump.p_out, ox_pump.p_out]) * gg.stiffness  # Pa - combustion chamber pressure guess
    gas_generator()

    # Turbine
    if engine.common_shaft == True:
        common_turbine = turbine()
        # TODO: power loss into bearings and seals
        common_turbine.shaft_efficiency = 1 # unitless - efficiency of power transmission through the shaft(s)
        common_turbine.shaft_power = (ox_pump.shaft_power+fuel_pump.shaft_power)/common_turbine.shaft_efficiency  
            # W - power that must be produced by the turbine
        common_turbine.shaft_speed = ox_pump.shaft_speed  # rad/s - both pumps on same shaft
        turbines(common_turbine)

    else:
        ox_turbine = turbine()
        fuel_turbine = turbine()
        ox_turbine.shaft_efficiency = turbine.gear_efficiency # unitless - efficiency of power transmission through the shaft(s)
        fuel_turbine.shaft_efficiency = turbine.gear_efficiency # unitless - efficiency of power transmission through the shaft(s)
        ox_turbine.shaft_power = ox_pump.shaft_power/ox_turbine.shaft_efficiency  # W - power that must be produced by the turbine
        fuel_turbine.shaft_power = fuel_pump.shaft_power/fuel_turbine.shaft_efficiency  # W - power that must be produced by the turbine
        turbines(turbine=ox_turbine)
        turbines(turbine=fuel_turbine)


    powerhead.ox_pump = ox_pump
    powerhead.fuel_pump = fuel_pump
    if engine.common_shaft == True:
        powerhead.turbine = common_turbine
    else:
        powerhead.ox_turbine = ox_turbine
        powerhead.fuel_turbine = fuel_turbine
